content/doc/book/pipeline/pipeline-best-practices.adoc

```diff
@@ -0,0 +1,95 @@
+
+Example:
+Instead of creating a series of  `echo` or `sh` steps, combine them into a single step or script.
+
+=== Avoid calls to `Jenkins.getInstance`
+
+Using Jenkins.instance or its accessor methods in a Pipeline or shared library indicates a code misuse within that Pipeline/shared library. Using Jenkins APIs from an unsandboxed shared library means that the shared library is both a shared library and a kind of Jenkins plugin. You need to be very careful when interacting with Jenkins APIs from a Pipeline to avoid severe security and performance issues. If you must use Jenkins APIs in your build, the recommended approach is to create a minimal plugin in Java that implements a safe wrapper around the Jenkins API you want to access using Pipeline's Step API. Using Jenkins APIs from a sandboxed Jenkinsfile directly means that you have probably had to whitelist methods that allow sandbox protections to be bypassed by anyone who can modify a Pipeline, which is a significant security risk. The whitelisted method is run as the System user, having overall admin permissions, which can lead to developers possessing higher permissions than intended.
+
+Solution: The best solution would be to work around the calls being made but if they must be done then it would be better to implement a Jenkins plugin which is able to gather the data needed.
+
+== Using shared libraries
+
+=== Do not override built-in Pipeline Steps
+
+Wherever possible stay away from customized/overwritten Pipeline steps. 
+Overriding built-in Pipeline Steps is the process of using shared libraries to overwrite the standard Pipeline APIs like `sh` or `timeout`. 
+This process is dangerous because the Pipeline APIs can change at any time causing custom code to break or give different results than expected. 
+When custom code breaks because of Pipeline API changes, troubleshooting is difficult because even if the custom code has not changed, it may not work the same after an API update.
+So even if custom code has not changed that does not mean that after an API update it will keep working the same. 
+Lastly, because of the ubiquitous use of these steps throughout Pipelines, if something is coded incorrectly/inefficiently the results can be catastrophic to Jenkins.
+
+=== Avoiding large global variable declaration files
+
+Having large variable declaration files can require large amounts of memory for little to no benefit, because the file is loaded for every Pipeline whether the variables are needed or not. Creating small variable files that contain only variables relevant to the current execution is recommended.
+It would be better to load variables which are relevant to the current execution.
```

Comment:
It would be better to only load variables which are relevant to the current execution.